quickwit/quickwit-common/src/pubsub.rs

```diff
@@ -0,0 +1,190 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex, Weak};
+use std::time::Duration;
+
+use async_trait::async_trait;
+
+use crate::type_map::TypeMap;
+
+pub trait Event: fmt::Debug + Clone + Send + Sync + 'static {}
+
+#[async_trait]
+pub trait EventSubscriber<E>: fmt::Debug + dyn_clone::DynClone + Send + Sync + 'static {
+    async fn handle_event(&mut self, event: E);
+}
+
+dyn_clone::clone_trait_object!(<E> EventSubscriber<E>);
+
+type EventSubscriptions<E> = HashMap<usize, EventSubscription<E>>;
+
+#[derive(Debug, Clone, Default)]
+pub struct EventBroker {
+    inner: Arc<InnerEventBroker>,
+}
+
+#[derive(Debug, Default)]
+struct InnerEventBroker {
+    subscription_sequence: AtomicUsize,
+    subscriptions: Mutex<TypeMap>,
+}
+
+impl EventBroker {
+    /// Subscribes to an event type.
+    pub fn subscribe<E>(&self, subscriber: impl EventSubscriber<E>) -> EventSubscriptionHandle<E>
+    where E: Event {
+        let mut subscriptions = self
+            .inner
+            .subscriptions
+            .lock()
+            .expect("The lock should never be poisoned.");
+
+        if !subscriptions.contains::<EventSubscriptions<E>>() {
+            subscriptions.insert::<EventSubscriptions<E>>(HashMap::new());
+        }
+        let subscription_id = self
+            .inner
+            .subscription_sequence
+            .fetch_add(1, Ordering::Relaxed);
+
+        let subscription = EventSubscription {
+            subscription_id,
+            subscriber: Box::new(subscriber),
+        };
+        let typed_subscriptions = subscriptions
+            .get_mut::<EventSubscriptions<E>>()
+            .expect("The subscription map should exist.");
+        typed_subscriptions.insert(subscription_id, subscription);
+
+        EventSubscriptionHandle {
+            subscription_id,
+            broker: Arc::downgrade(&self.inner),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Publishes an event.
+    pub fn publish<E>(&self, event: E)
+    where E: Event {
+        let subscriptions = self
+            .inner
+            .subscriptions
+            .lock()
+            .expect("The lock should never be poisoned.");
+
+        if let Some(typed_subscriptions) = subscriptions.get::<EventSubscriptions<E>>() {
+            for subscription in typed_subscriptions.values() {
+                let event = event.clone();
+                let mut subscriber = subscription.subscriber.clone();
+                tokio::spawn(tokio::time::timeout(Duration::from_secs(600), async move {
```

Comment:
don't know what would be a good timeout duration.

Nitpick: log a warning if we have a timeout
Nitpick: I would reduce the timeout but I don't have any good reasons, it's just that I would be scared if a handle_event can last 600 seconds...